### PR TITLE
issue#66 - Datasets Config

### LINKED ISSUE #66: AG dataset index import mechanism
Following up on the initial import mechanism from AG bucket to Elsa (#43 and #58), there are more things need to be done to completely import AG data to Elsa.

- [x] Mechanism on registering new Flagships. 

  Elsa would need to know each folder/prefix in S3 belongs to which dataset, by this there might be a mechanism of registering a map between prefix and dataset record. This process could also be the [initial creation](https://github.com/umccr/elsa-data/blob/26a597614fadbd9b53cbbe02da576c01f2d916c1/application/backend/src/api/routes/dataset-routes.ts#L83) of the dataset record for the files to be associated with.
 
- [x] Handle file deletion

	Some data might be requested to be deleted from AG bucket, ideally this should reflect on EdgeDb. Current implementation is again printing it to the console, probably this should be deleted and a popup warning shows the list of files to be deleted? Also to think, what happen to current grant access for the deleted file?
 
- [x] Complete the data model

	Some of the entries are trios (_mat / _pat) so these should go into a 'trio' case family group

- [ ] Get the Files to be included in SubmissionBatch objects that correspond to each folder

- [ ] Handle mysterious/incomplete submission.
  
  Some data may have pairs located in a different key-prefix/directory, ideally this would still be a valid a location but the current script would only consider pair from the same location. These skipped file is only printed in the console and need to be change so it could be inserted to EdgeDb appropriately.
  Note: If considering to find pair from different location, there is a chance that filename could be duplicated as part of reanalysis. This would be hard to identify which pair belongs to which. (Ref: https://trello.com/c/NFSr0Gjq)

- [ ] Decide on how/any phenotype information can come across into Elsa Data and where to source it

	Even basic things like "sex" and "age at sequencing" are not available to this Elsa Data import as it runs purely off S3 genotype information. What can/do we want to do here?

- [ ] DatasetPatient is no longer consented with the study. Current AG implementation is removing all related patient datas. 

	Do we also need to remove the patient record from `elsa`?  Perhaps another flag to show if patient is no longer available? A consent system exist in Elsa should this be enough without any other flag/deletion? 
	
	~ To be discussed with AG

### PR DESCRIPTION
**Pull request by williamputraintan**
- Add Configuration file and schema for elsa's dataset

### CODE REVIEW

File: application/backend/src/entrypoint.ts [hunk 1]
```diff
@@ -43,10 +44,20 @@ const start = async () => {
   container.register<ElsaSettings>("Settings", {
     useValue: settings,
   });
-
   await blankTestData();
   await insertTestData(settings);
 
+  // DB Setup
```
Comment (williamputraintan) [inline application/backend/src/entrypoint.ts]: Doubt on putting init dataset records here. I thought would be nice to have all datasets (in config) to be recorded or shown in UI despite of having an empty `DatasetCase`.

Possible to put these on `application/backend/src/entrypoint-command-start-web-server.ts` on [PR125](https://github.com/umccr/elsa-data/pull/125/files). Thoughts?

Comment (andrewpatto) [reply application/backend/src/entrypoint.ts]: Yes - this is a good idea - all the datasets in the config appearing in the UI.

Can we have some 'status' via the API dataset JSON that says whether the corresponding database record exists? (i.e. if noone has even attempted to do a 'load' for the dataset - maybe the UI can indicate this with a "warning" triangle icon or something)

Comment (williamputraintan) [reply application/backend/src/entrypoint.ts]: Do you mean the `DatasetCase`? Could do, yes. I think it is also explicit by seeing the _"0 artifacts ..."_ on the right side of the column in the datasets page.


Screen Shot 2022-11-24 at 10 20 39 am

Comment (andrewpatto) [reply application/backend/src/entrypoint.ts]: Not on DatasetCase - on the actual top level Dataset itself?

I'd contend there is a difference between a dataset that has been attempted to be loaded - and just has 0 artifacts currently - versus one that noone has yet attempted a load.  And 'isLoaded' is a boolean flag indicating whether the Dataset db record exists.

Also, we can add into Dataset a 'lastLoaded' /'lastRefreshed' datetime that is set every time your loader script is run?

File: application/backend/config/datasets.json5 [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (andrewpatto) [inline application/backend/config/datasets.json5:1]: I would think the datasets config that is actually checked into the application backend would just have our test data sets (10f etc).

A dataset config with cardiac will belong in the CDK deploy setup for UMCCR/AG

**Review by andrewpatto — APPROVED** [review]

File: application/backend/config/datasets.json5 (hunk 2 above)
Comment (andrewpatto) [reply application/backend/config/datasets.json5:1]: But happy if that is deferred until we move the CDK into its own repo

**Pull request merged**